src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+{
+    /// <summary>
+    /// Provider that computes display order of tree items based on input ordering of
+    /// evaluated includes from the project file.
+    /// </summary>
+    internal class TreeItemOrderPropertyProvider : IProjectTreePropertiesProvider
+    {
+        private const string FullPathProperty = "FullPath";
+        private readonly Dictionary<string, int> _displayOrderMap = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, int> _rootedOrderMap = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+        private readonly UnconfiguredProject _project;
+        private readonly ImmutableHashSet<string> _allowedItemTypes;
+
+        public TreeItemOrderPropertyProvider(IReadOnlyCollection<ProjectItemIdentity> orderedItems, UnconfiguredProject project)
+        {
+            _project = project;
+            OrderedItems = orderedItems;
+
+            _allowedItemTypes = OrderedItems.Select(p => p.ItemType).ToImmutableHashSet();
+
+            ComputeIndices();
+        }
+
+        public IReadOnlyCollection<ProjectItemIdentity> OrderedItems { get; }
+
+        /// <summary>
+        /// Preorder folders and items that are provided as ordered evaluated includes
+        /// </summary>
+        private void ComputeIndices()
+        {
+            var duplicateFiles = OrderedItems
+                .Select(p => Path.GetFileName(p.EvaluatedInclude))
+                .GroupBy(file => file, StringComparer.OrdinalIgnoreCase)
+                .Where(g => g.Count() > 1)
+                .Select(g => g.Key).ToImmutableHashSet();
+
+            var index = 1;
+
+            // This enumerates all the folder names and file names and maps them
+            // against their assigned display index. Any file names that are duplicates
+            // are mapped in in _rootedOrderMap only; everything else is mapped in
+            // _displayOrderMap only.
+            foreach (var item in OrderedItems)
+            {
+                var includeParts = item.EvaluatedInclude.Split(CommonConstants.PathSeparators, StringSplitOptions.RemoveEmptyEntries);
+                foreach (var part in includeParts)
```

Comment:
Yes, in the presence of duplicates especially involving folders, some of the display orders will be incorrect. This addresses some of the duplicate concerns, but should be replaced with a more comprehensive fix in CPS and Project System that resolves duplicates by providing more context for each item in the tree so it can be mapped to the EvaluatedInclude here. 

This current fix could be improved slightly by considering folders and filenames separately (and actually, all the folders in a given evaluated include could receive the same display order).

---
In reply to: [150064432](https://github.com/dotnet/project-system/pull/2930#discussion_r150064432) [](ancestors = 150064432)